recipe-client-addon/lib/StudyStorage.jsm

```diff
@@ -0,0 +1,210 @@
+// given value type.
+function validateSchema(data) {
+  const {schema} = DB_OPTIONS;
+
+  Object.keys(data).forEach((dataKey) => {
+    const incomingData = data[dataKey];
+    const requiredType = schema[dataKey] && schema[dataKey].type;
+
+    if (!requiredType) {
+      throw new Error(`Key "${dataKey}" does not exist in Study schema.`);
+    }
+
+    // In some cases, the value will be empty, at which point we just want to
+    // skip the rest of validation.
+    if (typeof incomingData === "undefined" || incomingData === null) {
+      return;
+    }
+
+    // If the `type` is a function, it's going to be a class (like Map or Set),
+    // otherwise we can assume it is a primitive type (like "boolean" or "number").
+    const validates = typeof requiredType === "function" ?
+      incomingData instanceof requiredType : typeof incomingData === requiredType;
+
+    if (!validates) {
+      throw new Error(`Key "${dataKey}" must be of type "${requiredType.name || requiredType}"`);
+    }
+  });
+}
+
+
+// Creates new object store and populates that store's indexes.
+function setupDatabase(db) {
+  const {schema} = DB_OPTIONS;
+  const store = db.createObjectStore(DB_NAME, { keyPath: "name" });
+
+  Object.keys(schema).forEach((schemaKey) => {
+    store.createIndex(schemaKey, schemaKey, schema[schemaKey]);
+  });
+}
+
+// Determines if a DB connection/request already exists.
+function hasDatabaseRequest() {
+  return !!gDatabase;
+}
+
+// Returns the currently active IDB connection, and creates one if not
+// already present.
+async function getDatabase() {
+  if (!hasDatabaseRequest()) {
+    // gDB is a Promise to prevent multiple calls to `IDB.open` - subsequent calls
+    // will refer to this same promise, requiring only one `open` call internally.
+    gDatabase = new Promise(async (resolve) => {
+      const db = await IndexedDB.open(DB_NAME, DB_OPTIONS, setupDatabase);
+      resolve(db);
+    });
+  }
+
+  return gDatabase;
+}
+
+
+this.StudyStorage = {
+  /**
+   * Clears all data from the connected IndexedDB.
+   * @async
+   */
+  async clearAllData() {
+    const db = await getDatabase();
+    await db.objectStore(DB_NAME, "readwrite").clear();
+  },
+
+  /**
+   * If a database exists and is connected, disconnect and remove the reference
+   * to that database.
+   *
+   * @async
+   */
+  async closeConnection() {
+    if (hasDatabaseRequest()) {
+      const db = await getDatabase();
+      await db.close();
+      gDatabase = null;
+    }
+  },
+
+  /**
+   * Look up a Study based on the given unique name. Returns `null` if no study
+   * is found in the DB.
+   *
+   * @async
+   * @param  {string} studyName Unique name of study to retrieve.
+   * @return {Promise<Study>}
+   */
+  async getStudy(studyName) {
+    const db = await getDatabase();
+
+    const foundStudy = await db.objectStore(DB_NAME, "readonly").get(studyName);
+
+    return foundStudy || null;
```

Comment:
What does the promise resolve to if the study is not present? The extra `|| null` feels odd - does it resolve to `false` ? Or `undefined` ? A comment might be helpful if it's really something other than `null` (which would kind of surprise me, but hey).